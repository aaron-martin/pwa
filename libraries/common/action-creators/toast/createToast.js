/**
 * Copyright (c) 2017 - present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { CREATE_TOAST } from '../../constants/ActionTypes';

let currentId = 0;
/**
 * Gets the next id.
 * @return {number}
 */
const getNextId = () => {
  currentId += 1;
  return currentId;
};

/**
 * Creates the dispatched CREATE_TOAST action object.
 * @param {Object} options The modal options.
 * @return {Object} A Redux action.
 */
const createToast = options => ({
  type: CREATE_TOAST,
  options: {
    ...options,
    id: getNextId(),
  },
});

export default createToast;
