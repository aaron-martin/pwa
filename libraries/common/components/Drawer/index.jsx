/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './style';

/**
 * Drawer component.
 */
class Drawer extends Component {
  /**
   * The component prop types.
   * @type {Object}
   */
  static propTypes = {
    alwaysActive: PropTypes.bool,
    animation: PropTypes.shape({
      duration: PropTypes.number,
      in: PropTypes.string,
      out: PropTypes.string,
    }),
    children: PropTypes.node,
    className: PropTypes.string,
    isOpen: PropTypes.bool,
    onClose: PropTypes.func,
    onDidClose: PropTypes.func,
    onOpen: PropTypes.func,
  };

  /**
   * The component default props.
   * @type {Object}
   */
  static defaultProps = {
    alwaysActive: false,
    className: '',
    children: null,
    isOpen: false,
    onOpen: () => {},
    onClose: () => {},
    onDidClose: () => {},
    animation: {
      duration: null,
      in: '',
      out: '',
    },
  };

  /**
   * Initializes the Drawer component.
   * @param {Object} props The components props.
   */
  constructor(props) {
    super(props);

    this.state = {
      active: props.isOpen,
    };
  }

  /**
   * Update state when isOpen changes.
   * @param {Object} nextProps The next component props.
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.isOpen !== nextProps.isOpen) {
      if (nextProps.isOpen) {
        if (typeof nextProps.onOpen === 'function') {
          nextProps.onOpen();
        }
        this.setState({ active: true });
      } else if (typeof nextProps.onClose === 'function') {
        nextProps.onClose();
      }
    }
  }

  /**
   * Syncs the internal active state when an animation ends.
   */
  handleAnimationEnd = () => {
    this.setState({ active: this.props.isOpen });
    if (!this.props.isOpen) {
      this.props.onDidClose();
    }
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      alwaysActive,
      className,
      children,
      isOpen,
      animation,
    } = this.props;
    const { active } = this.state;

    const animationIn = animation.in || styles.animation.in;
    const animationOut = animation.out || styles.animation.out;

    const combinedClassName = classNames(
      className,
      styles.container,
      { [animationIn]: isOpen },
      { [animationOut]: !isOpen }
    );

    const style = {};
    if (typeof animation.duration === 'number') {
      style.animationDuration = `${animation.duration}ms`;
    }

    return (active || alwaysActive) ? (
      <div
        className={combinedClassName}
        style={style}
        onAnimationEnd={this.handleAnimationEnd}
      >
        {children}
      </div>
    ) : null;
  }
}

export default Drawer;
