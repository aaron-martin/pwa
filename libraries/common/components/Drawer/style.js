/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

const duration = 150;
const easing = 'ease';

const slideInBaseDrawer = css.keyframes({
  '0%': { transform: 'translateY(100%)' },
  '100%': { transform: 'translateY(0)' },
});

const slideOutBaseDrawer = css.keyframes({
  '0%': { transform: 'translateY(0)' },
  '100%': { transform: 'translateY(100%)' },
});

const container = css({
  position: 'fixed',
  bottom: 0,
  left: 0,
  zIndex: 5,
}).toString();

const animation = {
  in: css({
    animation: `${slideInBaseDrawer} ${duration}ms 1 both ${easing}`,
  }).toString(),
  out: css({
    animation: `${slideOutBaseDrawer} ${duration}ms 1 both ${easing}`,
  }).toString(),
};

export default {
  container,
  animation,
};
