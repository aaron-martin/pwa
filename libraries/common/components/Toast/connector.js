/**
 * Copyright (c) 2017 - present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { connect } from 'react-redux';
import removeToast from '../../actions/toast/removeToast';
import { getToast, hasNextToast } from '../../selectors/toast';

/**
 * Maps state to props.
 * @param {Object} state State.
 * @returns {Object}
 */
const mapStateToProps = state => ({
  toast: getToast(state),
  hasNextToast: hasNextToast(state),
});

/**
 * Maps dispatch to props.
 * @param {function} dispatch Dispatch.
 * @returns {Object}
 */
const mapDispatchToProps = dispatch => ({
  removeToast: (id) => {
    dispatch(removeToast(id));
  },
  // Used for dispatching prepared action stored in a toast message.
  dispatchAction: action => dispatch(action),
});

export default connect(mapStateToProps, mapDispatchToProps);
