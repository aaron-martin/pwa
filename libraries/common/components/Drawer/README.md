# Drawer
---
This component transitions based on it's `isOpen` property. It is intended to be used to provide an additional UI layer to a page.

## Getting Started

```markup
import { Drawer } from 'Library/components';

<Drawer isOpen={true}>
  <SomeContent />
</Drawer>
```

## Props

### className

_Default_: `''`
_Type_: `string`

Additional classes that can be appended to the component's `className` attribute.

###### Usage:
```
<Drawer className="class1 class2">
  ...
</Drawer>
```

### children

_Default_: `null`
_Type_: `node`

Children that will be rendered.

###### Usage:

```
<Drawer>
  <ChildComponent />
</Drawer>
```

### isOpen

_Default_: `false`
_Type_: `boolean`

Controls whether or not the drawer is open.

###### Usage:

```
<Drawer isOpen={true}>
  ...
</Drawer>
```

### onOpen

_Default_: `() => {}`
_Type_: `Function`

A callback function that is fired after the Drawer has been opened.

###### Usage:

```
const myFunc = () => {
  ...
};

<Drawer onOpen={myFunc}>
  ...
</Drawer>
```

### onClose

_Default_: `() => {}`
_Type_: `Function`

A callback function that is fired after the Drawer has been closed.

###### Usage:

```
const myFunc = () => {
  ...
};

<Drawer onClose={myFunc}>
  ...
</Drawer>
```

### animation

_Default_:

```
{
  duration: 300,
  easing: 'ease-out',
  from: null,
  set: null,
  to: null,
}
```
_Type_: `Object`

An object that describes the transition of the Drawer.

###### Usage:

```
const drawerAnimation = {
  from: {
    y: '100%'
  },
  to: {
    y: '0%'
  },
};

<Drawer animation={drawerAnimation}>
  ...
</Drawer>
```
