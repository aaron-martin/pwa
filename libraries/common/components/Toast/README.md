# Toast
---

Toast components which is responsible for maintaining the toast messages and rendering them by using a provided compoments.

The main idea behind this component is to just provide the logic behind showing and hiding Toasts. Rendering of all elements like text and even the outer container is delegated to components which are given via props.

## Getting Started

```jsx
import { Toast } from 'Library/components/Toast';

<Toast message={Message} container={Container} />
```

When `Toast` is rendered it would always open a `Drawer`. `Drawer` would contain `Container` and rendered text in a `Message` component.

The end result would be similar to this:
```jsx
<Toast>
    <div aria-live="polite" aria-relevant="additions">
        <Drawer>
            <Container>
                <Message text={'string'} />
            </Container>
        </Drawer>
    </div>
</Toast>
``` 

When the message times out, `Drawer` is closed and message is removed from redux store.

If there are more messages in redux store, the messages would be shown one by one consecutively.

## How to show a toast message.
When `Toast` component is rendered in the app, it waits until there's a message added to a redux store.
In order to add this message, simply use an action available in `Library/actions/toast/createToast`.

## How to remove a toast message.
Toast messages should not be removed from the outside. Removed message which is a subject of current visible toast message would suddenly disappear. This component handles messages removal in a clean way.

## How to show action with a callback
Toast messages can have single action button with a callback. When the button is clicked, callback would be executed and toast would hide.

To show a message with action button simply use an action available in `Library/actions/toast/createToast` with options as follows:
```js
createToast({
  message: 'Example message',
  action: 'Click me!',
  actionOnClick: () => alert('Clicked!')
});
```
## Styling
Styling is mostly done by the `.message` and `.container` props. You can however pass `.className` prop which would be used
by a `Drawer`.

## Props

### message (required)

_Type_: `React.Component`

Component which would render the message text. It must accept and render `.text` prop. 

### container (required)

_Type_: `React.Component`

Component which would render the `Message` component. The only requirement to this component is that it has to render children.

### className (optional)

_Type_: `string`

Class name which is passed to a `Drawer` component and affect the `Drawer` styling.

### onClose (optional)

_Type_: `function`

Function which would be called every time the Drawer will close. 

## Accessibility
The `Toast` component is wrapped with a `aria-live` container. It tells the screen readers to read every incoming message in a `polite` mode.

