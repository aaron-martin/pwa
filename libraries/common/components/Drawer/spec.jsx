/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Drawer from './index';

describe('<Drawer />', () => {
  let mockOpen;

  beforeEach(() => {
    mockOpen = jest.fn();
  });

  it('should render', () => {
    const wrapper = shallow(<Drawer />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should execute callback when drawer is opened', () => {
    const wrapper = mount(<Drawer onOpen={mockOpen} />);
    wrapper.setProps({
      isOpen: true,
    });

    expect(mockOpen).toBeCalled();
  });

  it('should execute callback when drawer is closed', () => {
    const wrapper = mount(<Drawer isOpen onClose={mockOpen} />);
    wrapper.setProps({
      isOpen: false,
    });

    expect(mockOpen).toBeCalled();
  });

  it('should add custom classes', () => {
    const wrapper = mount(<Drawer className="custom-class-name" isOpen />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('custom-class-name')).toEqual(true);
  });

  it('should execute callback when drawer close animation did end', () => {
    const wrapper = mount(<Drawer className="custom-class-name" isOpen onDidClose={mockOpen} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({
      isOpen: false,
    });
    wrapper.simulate('animationEnd');
    expect(mockOpen).toBeCalled();
  });
});
