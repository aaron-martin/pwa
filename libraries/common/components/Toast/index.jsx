/**
 *  Copyright (c) 2017 - present, Shopgate, Inc. All rights reserved.
 *
 *  This source code is licensed under the Apache 2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Drawer from '../Drawer';
import connect from './connector';

/**
 * Toast component.
 */
class Toast extends Component {
  static propTypes = {
    container: PropTypes.func.isRequired,
    message: PropTypes.func.isRequired,
    removeToast: PropTypes.func.isRequired,
    className: PropTypes.string,
    onClose: PropTypes.func,
    toast: PropTypes.shape(),
  };

  static defaultProps = {
    className: null,
    onClose: () => {},
    toast: null,
  };

  /**
   * Constructs.
   * @param {Object} props Props.
   */
  constructor(props) {
    super(props);
    this.state = {
      isOpen: !!props.toast,
    };
    this.timeout = null;
  }

  /**
   * ComponentWillReceiveProps.
   * @param {Object} nextProps NextProps.
   */
  componentWillReceiveProps(nextProps) {
    const hasToast = !!nextProps.toast;

    if (hasToast) {
      this.handleTimeout(nextProps.toast);
    }

    if (!this.state.isOpen) {
      this.setState({
        isOpen: hasToast,
      });
    }
  }

  /**
   * ShouldComponentUpdate.
   * @param {Object} nextProps NextProps.
   * @param {Object} nextState NextState.
   * @return {boolean}
   */
  shouldComponentUpdate(nextProps, nextState) {
    const toastUpdate = nextProps.toast && this.props.toast
      && nextProps.toast.id !== this.props.toast.id;
    const toastDidChange = toastUpdate || !this.props.toast;

    if (nextState.isOpen !== this.state.isOpen) {
      return true;
    }

    return toastDidChange;
  }

  /**
   * Closes Drawer or removes the message when drawer is closed.
   */
  handleRemoveMessage = () => {
    if (this.props.toast) {
      this.props.removeToast(this.props.toast.id);
    }
  };

  /**
   * Handles timeout.
   * @param {Object} toast Toast.
   */
  handleTimeout = (toast) => {
    if (toast.duration) {
      clearTimeout(this.timeout);
      this.timeout = setTimeout(() => {
        this.setState({
          isOpen: false,
        });
        this.props.onClose();
      }, toast.duration);
    }
  };

  /**
   * Renders.
   * @returns {XML}
   */
  render() {
    const Container = this.props.container;
    const Message = this.props.message;

    return (
      <div
        aria-live="polite"
        aria-relevant="additions"
      >
        <Drawer
          isOpen={this.state.isOpen}
          onDidClose={this.handleRemoveMessage}
          className={this.props.className}
        >
          {
            this.props.toast &&
            <Container>
              <Message text={this.props.toast.message} />
            </Container>
          }
        </Drawer>
      </div>
    );
  }
}

export default connect(Toast);
