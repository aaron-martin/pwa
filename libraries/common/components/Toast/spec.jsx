/**
 * Copyright (c) 2017 - present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import Drawer from '../Drawer';
import Toast from './index';
import configureStore from '../../store';
import toastReducer from '../../reducers/toast';
import createToast from '../../actions/toast/createToast';
import removeToast from '../../actions/toast/removeToast';
import { getToast } from '../../selectors/toast';
import mockRenderOptions from '../../helpers/mocks/mockRenderOptions';

const mockedStore = configureStore({ toast: toastReducer });

/* eslint-disable react/prop-types */
/**
 * Container Mock
 * @param {Object} props Props
 * @returns {JSX}
 */
const MockContainer = props => (<div>{props.children}</div>);

/**
 * Message Mock
 * @param {string} text Message text
 * @returns {JSX}
 */
const MockMessage = ({ text }) => (<span>{text}</span>);

/**
 * Action Button.
 * @param {string} text Action text.
 * @param {function} onClick OnClick callback.
 * @returns {JSX}
 */
const MockActionButton = ({ text, onClick }) => (<button onClick={onClick} >{text}</button>);
/* eslint-enable react/prop-types */

/**
 * Creates the component
 * @param {function} onClose Close callback
 * @return {*} ReactWrapper
 */
const createComponent = (onClose) => {
  /* eslint-disable global-require */
  const Toast = require('./index').default; // eslint-disable-line no-shadow
  /* eslint-enable global-require */
  return mount(
    <Provider store={mockedStore}>
      <Toast
        container={MockContainer}
        message={MockMessage}
        actionButton={MockActionButton}
        toast={getToast}
        onClose={onClose}
      />
    </Provider>,
    mockRenderOptions
  );
};

jest.mock('redux-logger', () => ({
  /**
   * Create logger
   * @returns {*} action
   */
  createLogger: () => () => next => action => next(action),
}));

describe('<Toast />', () => {
  const { dispatch, getState } = mockedStore;
  let wrapper;

  beforeEach(() => {
    // Reset the toasts state before each test.
    getState().toast = [];
    wrapper = createComponent();
  });

  it('should dispatch a toast message', () => {
    dispatch(createToast({ message: 'Toast Message', duration: 0 }));
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockMessage).exists()).toBe(true);
    expect(wrapper.find(MockContainer).exists()).toBe(true);
  });

  it('should remove after timeout', (done) => {
    const wrapperWithTimeout = createComponent(() => {
      const drawerCallback = wrapperWithTimeout.find(Drawer).prop('onDidClose');
      drawerCallback();
      expect(getState().toast.length).toBe(0);
      done();
    });
    dispatch(createToast({ message: 'Timeout Message', duration: 10 }));
    wrapperWithTimeout.update();
  });

  it('should dispatch multiple toast messages', () => {
    const messages = ['Toast Message 2', 'Toast Message 3'];
    const getNextToast = wrapper.find(Toast).prop('toast');

    messages.forEach((message) => {
      dispatch(createToast({ message, duration: 0 }));
    });
    wrapper.update();

    const msg1 = getNextToast(getState());
    expect(wrapper.find(MockMessage).prop('text')).toEqual(msg1.message);
    dispatch(removeToast(msg1.id));
    wrapper.update();

    const msg2 = getNextToast(getState());
    expect(wrapper.find(MockMessage).prop('text')).toEqual(msg2.message);
  });
  it('should call actionOnClick on action click', (done) => {
    const mockedActionOnClick = jest.fn();
    dispatch(createToast({
      message: 'hello world',
      action: 'click me',
      actionOnClick: mockedActionOnClick,
      duration: 2,
    }));
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockMessage).exists()).toBe(true);
    expect(wrapper.find(MockActionButton).exists()).toBe(true);
    wrapper.find(MockActionButton).simulate('click');
    wrapper.update();
    setTimeout(() => {
      wrapper.update();
      expect(typeof mockedActionOnClick.mock.calls[0][0] === 'function').toBe(true);
      expect(typeof mockedActionOnClick.mock.calls[0][1] === 'function').toBe(true);
      expect(wrapper.find(MockMessage).exists()).toBe(false);
      expect(wrapper).toMatchSnapshot();
      done();
    }, 0);
  });
});
