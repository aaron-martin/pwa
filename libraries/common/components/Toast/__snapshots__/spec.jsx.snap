// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toast /> should dispatch a toast message 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Toast)
    container={[Function]}
    message={[Function]}
    toast={[Function]}
  >
    <Toast
      className={null}
      container={[Function]}
      message={[Function]}
      onClose={[Function]}
      removeToast={[Function]}
      toast={
        Object {
          "duration": 0,
          "id": 1,
          "message": "Toast Message",
        }
      }
    >
      <div
        aria-live="polite"
        aria-relevant="additions"
      >
        <Drawer
          alwaysActive={false}
          animation={
            Object {
              "duration": null,
              "in": "",
              "out": "",
            }
          }
          className={null}
          isOpen={true}
          onClose={[Function]}
          onDidClose={[Function]}
          onOpen={[Function]}
        >
          <div
            className="css-1rg7udn css-1dwe0qx"
            onAnimationEnd={[Function]}
            style={Object {}}
          >
            <MockContainer>
              <div>
                <MockMessage
                  text="Toast Message"
                >
                  <span>
                    Toast Message
                  </span>
                </MockMessage>
              </div>
            </MockContainer>
          </div>
        </Drawer>
      </div>
    </Toast>
  </Connect(Toast)>
</Provider>
`;
